File: building-reusable-abstractions/2-component-resources/go/main.go
package main

import (
	"github.com/pulumi/pulumi-civo/sdk/go/civo"
	k8s "github.com/pulumi/pulumi-kubernetes/sdk/v3/go/kubernetes"
	appsv1 "github.com/pulumi/pulumi-kubernetes/sdk/v3/go/kubernetes/apps/v1"
	corev1 "github.com/pulumi/pulumi-kubernetes/sdk/v3/go/kubernetes/core/v1"
	metav1 "github.com/pulumi/pulumi-kubernetes/sdk/v3/go/kubernetes/meta/v1"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

func main() {
	pulumi.Run(func(ctx *pulumi.Context) error {
		network, err := civo.NewNetwork(ctx, "network-two", &civo.NetworkArgs{
			Label: pulumi.String("pulumi-workshop"),
		})

		if err != nil {
			return err
		}

		cluster, err := civo.NewKubernetesCluster(ctx, "cluster", &civo.KubernetesClusterArgs{
			Name:           pulumi.String("cluster"),
			NetworkId:      network.ID(),
			NumTargetNodes: pulumi.Int(2),
			Applications:   pulumi.String("metrics-server,-Traefik-v2-nodeport"),
		})

		if err != nil {
			return err
		}

		kubernetesProvider, err := k8s.NewProvider(ctx, "kubernetesProvider", &k8s.ProviderArgs{
			Kubeconfig: cluster.Kubeconfig,
		})

		if err != nil {
			return err
		}

		_, err = appsv1.NewDeployment(ctx, "nginx", &appsv1.DeploymentArgs{
			Metadata: &metav1.ObjectMetaArgs{
				Labels: pulumi.StringMap{
					"app": pulumi.String("nginx"),
				},
			},
			Spec: &appsv1.DeploymentSpecArgs{
				Replicas: pulumi.Int(3),
				Selector: &metav1.LabelSelectorArgs{
					MatchLabels: pulumi.StringMap{
						"app": pulumi.String("nginx"),
					},
				},
				Template: &corev1.PodTemplateSpecArgs{
					Metadata: &metav1.ObjectMetaArgs{
						Labels: pulumi.StringMap{
							"app": pulumi.String("nginx"),
						},
					},
					Spec: &corev1.PodSpecArgs{
						Containers: corev1.ContainerArray{
							&corev1.ContainerArgs{
								Name:  pulumi.String("nginx"),
								Image: pulumi.String("nginx:1.14.2"),
								Ports: corev1.ContainerPortArray{
									&corev1.ContainerPortArgs{
										ContainerPort: pulumi.Int(80),
									},
								},
							},
						},
					},
				},
			},
		}, pulumi.Provider(kubernetesProvider))

		if err != nil {
			return err
		}

		service, err := corev1.NewService(ctx, "nginx", &corev1.ServiceArgs{
			Spec: &corev1.ServiceSpecArgs{
				Type: pulumi.String("NodePort"),
				Selector: pulumi.StringMap{
					"app": pulumi.String("nginx"),
				},
				Ports: corev1.ServicePortArray{
					&corev1.ServicePortArgs{
						Protocol: pulumi.String("TCP"),
						Port:     pulumi.Int(80),
					},
				},
			},
		}, pulumi.Provider(kubernetesProvider))

		if err != nil {
			return err
		}

		nginxUrl := pulumi.All(cluster.MasterIp, service.Spec.Ports()).ApplyT(func(args []interface{}) pulumi.StringOutput {
			clusterIp := args[0].(string)
			ports := args[1].([]corev1.ServicePort)

			return pulumi.Sprintf("http://%s:%d", clusterIp, *ports[0].NodePort)
		})

		ctx.Export("nginxUrl", nginxUrl)
		ctx.Export("kubeconfig", cluster.Kubeconfig)

		return nil
	})
}

File: building-reusable-abstractions/2-component-resources/python/__main__.py
import json
import pulumi
import pulumi_civo as civo
import pulumi_kubernetes as kubernetes

network = civo.Network("network-two", label="pulumi-workshop")

firewall = civo.Firewall("firewall-two", network_id=network.id)

cluster = civo.KubernetesCluster('cluster', name='cluster',
                                 num_target_nodes=2, target_nodes_size='g3.k3s.medium',
                                 network_id=network.id, firewall_id=firewall.id)

kubernetes_provider = kubernetes.Provider(
    "kubernetes-provider", kubeconfig=cluster.kubeconfig)

kubernetes.apps.v1.Deployment(
    "nginx",
    metadata=kubernetes.meta.v1.ObjectMetaArgs(
        labels={
            "app": "nginx",
        },
    ),
    spec=kubernetes.apps.v1.DeploymentSpecArgs(
        replicas=3,
        selector=kubernetes.meta.v1.LabelSelectorArgs(
            match_labels={
                "app": "nginx",
            },
        ),
        template=kubernetes.core.v1.PodTemplateSpecArgs(
            metadata=kubernetes.meta.v1.ObjectMetaArgs(
                labels={
                    "app": "nginx",
                },
            ),
            spec=kubernetes.core.v1.PodSpecArgs(
                containers=[kubernetes.core.v1.ContainerArgs(
                    name="nginx",
                    image="nginx:1.14.2",
                    ports=[kubernetes.core.v1.ContainerPortArgs(
                        container_port=80,
                    )],
                )],
            ),
        ),
    ), opts=pulumi.ResourceOptions(provider=kubernetes_provider))

service = kubernetes.core.v1.Service(
    "nginx",
    spec=kubernetes.core.v1.ServiceSpecArgs(
        selector={
            "app": "nginx",
        },
        type="NodePort",
        ports=[kubernetes.core.v1.ServicePortArgs(
            protocol="TCP",
            port=80,
        )],
    ), opts=pulumi.ResourceOptions(provider=kubernetes_provider))

nginx_url = pulumi.Output.all(cluster.master_ip, service.spec.ports).apply(
    lambda args: pulumi.Output.concat("http://", args[0], ":", str(args[1][0]['node_port'])))

pulumi.export("nginxUrl", nginx_url)
pulumi.export("kubeconfig", cluster.kubeconfig)
